File: C++/even_odd_tree.cpp
// no,1609
#include "tree_node.h"
#include <vector>
using namespace std;

class Solution {
public:
  bool validateEvenRow(vector<int> vec) {
    if (vec.back() % 2 == 0) {
      return false;
    }
    for (int i = 0; i < vec.size() - 1; i++) {
      if ((vec[i] >= vec[i + 1]) | (!vec[i] % 2)) {
        return false;
      }
    }
    return true;
  }
  bool validateOddRow(vector<int> vec) {
    if (vec.back() % 2 == 1) {
      return false;
    }
    for (int i = 0; i < vec.size() - 1; i++) {
      if ((vec[i] <= vec[i + 1]) | (vec[i] % 2)) {
        return false;
      }
    }
    return true;
  }
  bool isEvenOddTree(TreeNode *root) {
    vector<TreeNode *> queue;
    queue.push_back(root);
    vector<int> row;
    vector<TreeNode *> next_row;
    bool ans;
    int counter = 0;
    while (!queue.empty()) {
      row.clear();
      next_row.clear();
      for (TreeNode *node : queue) {
        row.push_back(node->val);
        if (node->left) {
          next_row.push_back(node->left);
        }
        if (node->right) {
          next_row.push_back(node->right);
        }
      }
      if (counter % 2) {
        ans = validateOddRow(row);
      } else {
        ans = validateEvenRow(row);
      }
      if (!ans) {
        return false;
      }
      queue = next_row;
      counter++;
    }
    return true;
  }
};

File: python/even_odd_tree.py
#no. 1609
from utils import TreeNode

class Solution:
    def validate_even_row(self, vec: List[int]):
        if vec[-1] % 2 == 0:
            return False
        for i in range(len(vec) - 1):
            if(vec[i] >= vec[i+1]) or (vec[i] % 2 == 0):
                return False
        return True

    def validate_odd_row(self, vec: List[int]):
        if vec[-1] % 2 == 1:
            return False
        for i in range(len(vec) - 1):
            if(vec[i] <= vec[i+1]) or (vec[i] % 2 == 1):
                return False
        return True

    def isEvenOddTree(self, root: Optional[TreeNode]) -> bool:
        queue = [root]
        counter = 0
        while queue:
            row = []
            next_row = []
            for node in queue:
                row.append(node.val)
                if node.left:
                    next_row.append(node.left)
                if node.right:
                    next_row.append(node.right)
            if counter % 2: # odd
                ans = self.validate_odd_row(row)
            else: # even
                ans = self.validate_even_row(row)
            if not ans:
                return False
            queue = next_row
            counter += 1
        return True
